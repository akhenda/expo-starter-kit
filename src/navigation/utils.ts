import React, { useCallback, useEffect, useLayoutEffect, useRef, useState } from 'react';
import { BackHandler } from 'react-native';
import { createNavigationContainerRef, NavigationState, useNavigation } from '@react-navigation/native';

import type { PersistNavigationConfig } from '@config/config.base';
import Config from '@src/config';
import { useIsMounted } from '@src/hooks/useLifecycle';
import { mmkvStorage as storage } from '@src/services/storage';

import type { RootStackParamList } from './navigators/RootNavigator';

export const NAVIGATION_PERSISTENCE_KEY = '@NAVIGATION_STATE';

export const navigationRef = createNavigationContainerRef<RootStackParamList>();

/**
 * Gets the current screen from any navigation state.
 */
export function getActiveRouteName(state?: NavigationState<RootStackParamList>): keyof RootStackParamList {
  if (!state || !state.index) return 'App Stack';

  const route = state.routes[state.index];

  if (!route) return 'App Stack';

  // Found the active route -- return the name
  if (!route.state) return route.name;

  // Recursive call to deal with nested routers
  return getActiveRouteName(route.state as NavigationState<RootStackParamList>);
}

/**
 * Hook that handles Android back button presses and forwards those on to
 * the navigation or allows exiting the app.
 */
export function useBackButtonHandler(canExit: (routeName: keyof RootStackParamList) => boolean) {
  // The reason we're using a ref here is because we need to be able
  // to update the canExit function without re-setting up all the listeners
  const canExitRef = useRef(canExit);

  useEffect(() => {
    canExitRef.current = canExit;
  }, [canExit]);

  useEffect(() => {
    // We'll fire this when the back button is pressed on Android.
    const onBackPress = () => {
      if (!navigationRef.isReady()) return false;

      // grab the current route
      const routeName = getActiveRouteName(navigationRef.getRootState() as NavigationState<RootStackParamList>);

      // are we allowed to exit?
      if (canExitRef.current(routeName)) {
        // exit and let the system know we've handled the event
        BackHandler.exitApp();

        return true;
      }

      // we can't exit, so let's turn this into a back action
      if (navigationRef.canGoBack()) {
        navigationRef.goBack();

        return true;
      }

      return false;
    };

    // Subscribe when we come to life
    BackHandler.addEventListener('hardwareBackPress', onBackPress);

    // Unsubscribe when we're done
    return () => BackHandler.removeEventListener('hardwareBackPress', onBackPress);
  }, []);
}

/**
 * This helper function will determine whether we should enable navigation persistence
 * based on a config setting and the __DEV__ environment (dev or prod).
 */
function navigationRestoredDefaultState(persistNavigation: PersistNavigationConfig) {
  if (persistNavigation === 'always') return false;
  if (persistNavigation === 'dev' && __DEV__) return false;
  if (persistNavigation === 'prod' && !__DEV__) return false;

  // all other cases, disable restoration by returning true
  return true;
}

/**
 * Custom hook for persisting navigation state.
 */
export function useNavigationPersistence(persistenceKey: string = NAVIGATION_PERSISTENCE_KEY) {
  const isMounted = useIsMounted();
  const routeNameRef = useRef<string | undefined>();
  const initNavState = navigationRestoredDefaultState(Config.persistNavigation);

  const [initialNavigationState, setInitialNavigationState] = useState<NavigationState<RootStackParamList>>();
  const [isRestored, setIsRestored] = useState(initNavState);

  const onNavigationStateChange = (state?: NavigationState) => {
    const previousRouteName = routeNameRef.current;
    const currentRouteName = getActiveRouteName(state as NavigationState<RootStackParamList>);

    if (previousRouteName !== currentRouteName) {
      // track screens.
      if (__DEV__)
        // eslint-disable-next-line no-console
        console.tron.display?.({
          name: 'currentRouteName',
          preview: currentRouteName,
          value: currentRouteName,
        });
    }

    // Save the current route name for later comparison
    routeNameRef.current = currentRouteName;

    // Persist state to storage
    storage.save(persistenceKey, state);
  };

  const restoreState = useCallback(async () => {
    try {
      const state = (await storage.load(persistenceKey)) as NavigationState<RootStackParamList>;

      if (state) setInitialNavigationState(state);
    } finally {
      if (isMounted()) setIsRestored(true);
    }
  }, [persistenceKey, isMounted]);

  useEffect(() => {
    if (!isRestored) restoreState();
  }, [isRestored, restoreState]);

  return { initialNavigationState, isRestored, onNavigationStateChange, restoreState };
}

/**
 * A hook that can be used to easily set the Header of a react-navigation screen from within the screen's component.
 *
 * - [Documentation and Examples](https://github.com/infinitered/ignite/blob/master/docs/Utils-useHeader.md)
 */
export function useHeader(header: React.ReactNode, deps: unknown[] = []) {
  const navigation = useNavigation();

  useLayoutEffect(() => {
    navigation.setOptions({
      header,
      headerShown: true,
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [...deps, header, navigation]);
}

/**
 * use this to navigate without the navigation
 * prop. If you have access to the navigation prop, do not use this.
 * More info: https://reactnavigation.org/docs/navigating-without-navigation-prop/
 */
export function navigate(...args: Parameters<typeof navigationRef.navigate>): void {
  if (navigationRef.isReady()) navigationRef.navigate(...args);
}

export function goBack() {
  if (navigationRef.isReady() && navigationRef.canGoBack()) navigationRef.goBack();
}

export function resetRoot(params = { index: 0, routes: [] }) {
  if (navigationRef.isReady()) navigationRef.resetRoot(params);
}
