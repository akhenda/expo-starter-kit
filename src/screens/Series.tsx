import { useState } from 'react';
import { ScrollView } from 'react-native';
import Button from 'react-native-ui-lib/button';
import Text from 'react-native-ui-lib/text';
import TextField from 'react-native-ui-lib/textField';
import View from 'react-native-ui-lib/view';
import isEqual from 'lodash/isEqual';

import { SafeAreaView, SaveAreaInset } from '@src/components';
import { PaletteIcon } from '@src/components/Icons';
import logger from '@src/services/logger';
import { actions, useTrackedStore } from '@src/stores';

type ButtonProps = {
  label?: string;
  round?: boolean;
};

const labelButton = { label: 'Change Theme' };
const iconButton = { round: true };

const Series = () => {
  const theme = useTrackedStore().app.theme();
  const [email, setEmail] = useState('');
  const [buttonProps, setButtonProps] = useState<ButtonProps>(labelButton);

  const onChangeTheme = () => {
    if (isEqual(buttonProps, labelButton)) setButtonProps(iconButton);
    if (isEqual(buttonProps, iconButton)) setButtonProps(labelButton);

    actions.app.changeTheme(theme === 'dark' ? 'light' : 'dark');

    logger.info('Changing theme in Series');
  };

  const text = 'Series';

  return (
    <SafeAreaView insets={SaveAreaInset.BOTTOM}>
      <ScrollView>
        <View flex padding-15>
          <Text h1 text>
            H1 {text}
          </Text>
          <Text h2 secondary>
            H2 {text}
          </Text>
          <Text h3 success>
            H3 {text}
          </Text>
          <Text h4 warning>
            H4 {text}
          </Text>
          <Text h5 info>
            H5 {text}
          </Text>
          <Text h6 error>
            H6 {text}
          </Text>
          <Text caption>Caption</Text>
          <Text overline>Overline</Text>
          <Text subtitle>Subtitle</Text>
          <Text subtitle2>Subtitle 2</Text>
          <Text body>Body</Text>
          <Text testID="email" body2>
            {email}
          </Text>
          <Text testID="theme" p>
            {theme}
          </Text>
          <TextField
            placeholder="Placeholder"
            onChangeText={setEmail}
            enableErrors
            validate={['required', 'email', (value: string) => value.length > 6]}
            validationMessage={['Field is required', 'Email is invalid', 'Password is too short']}
            showCharCounter
            maxLength={30}
          />
          <Button
            testID="button"
            onPress={onChangeTheme}
            marginT-15
            bg-primary
            enableShadow
            iconSource={PaletteIcon}
            animateLayout
            iconOnRight
            animateTo="right"
            {...buttonProps}
          />
        </View>
      </ScrollView>
    </SafeAreaView>
  );
};

export default Series;
