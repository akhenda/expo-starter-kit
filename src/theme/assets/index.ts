export { default as images } from './images';
export { default as lottie } from './lottie';
