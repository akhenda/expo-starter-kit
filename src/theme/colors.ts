/* eslint-disable sort-keys-fix/sort-keys-fix */
/**
 * Stargate Color Pallete: #0099cc,#194351,#29576e,#4285b0,#1c7ab0
 *
 * @see https://coolors.co/0099cc-194351-29576e-4285b0-1c7ab0
 * @see https://www.color-hex.com/color-palette/89198
 */
const stargateColors = {
  blueGreen: '#0099cc',
  midnightGreenEagleGreen: '#194351',
  blueSapphire: '#29576e',
  steelBlue: '#4285b0',
  starCommandBlue: '#1c7ab0',
} as const;

/**
 * https://colors.eva.design/
 */
const palette = {
  primary100: '#D0F6FB',
  primary200: '#A3E8F7',
  primary300: '#72CBE7',
  primary400: '#4CA7CF',
  primary500: '#1C7AB0',
  primary600: '#145F97',
  primary700: '#0E477E',
  primary800: '#083266',
  primary900: '#052454',

  success100: '#EBFBD0',
  success200: '#D3F7A2',
  success300: '#AEE771',
  success400: '#88D04B',
  success500: '#57B21A',
  success600: '#409913',
  success700: '#2D800D',
  success800: '#1D6708',
  success900: '#125504',

  info100: '#D3E4FF',
  info200: '#A8C8FF',
  info300: '#7CA8FF',
  info400: '#5C8EFF',
  info500: '#2663ff',
  info600: '#1B4CDB',
  info700: '#1337B7',
  info800: '#0C2693',
  info900: '#071A7A',

  warning100: '#FEF7CB',
  warning200: '#FEEE98',
  warning300: '#FEE165',
  warning400: '#FDD53F',
  warning500: '#fcc100',
  warning600: '#D8A000',
  warning700: '#B58200',
  warning800: '#926500',
  warning900: '#785000',

  danger100: '#FDE7D0',
  danger200: '#FBC9A1',
  danger300: '#F4A271',
  danger400: '#E97C4D',
  danger500: '#db4518',
  danger600: '#BC2C11',
  danger700: '#9D180C',
  danger800: '#7F0907',
  danger900: '#69040A',

  overlay20: 'rgba(25, 16, 21, 0.2)',
  overlay50: 'rgba(25, 16, 21, 0.5)',
} as const;

/**
 * https://bareynol.github.io/mui-theme-creator
 */
const derived = {
  primary: {
    main: palette.primary500,
    light: 'rgb(73, 148, 191)',
    dark: 'rgb(19, 85, 123)',
    contrastText: '#fff',
  },
  secondary: {
    main: palette.success500,
    light: 'rgb(120, 193, 71)',
    dark: 'rgb(60, 124, 18)',
    contrastText: 'rgba(0, 0, 0, 0.87)',
  },
  error: {
    main: palette.danger500,
    light: 'rgb(226, 106, 70)',
    dark: 'rgb(153, 48, 16)',
    contrastText: 'fff',
  },
  warning: {
    main: palette.warning500,
    light: 'rgb(252, 205, 51)',
    dark: 'rgb(176, 135, 0)',
    contrastText: 'rgba(0, 0, 0, 0.87)',
  },
  info: {
    main: palette.info500,
    light: 'rgb(81, 130, 255)',
    dark: 'rgb(26, 69, 178)',
    contrastText: '#fff',
  },
  success: {
    main: palette.success500,
    light: 'rgb(120, 193, 71)',
    dark: 'rgb(60, 124, 18)',
    contrastText: 'rgba(0, 0, 0, 0.87)',
  },

  text: {
    dark: {
      primary: '#fff',
      secondary: 'rgba(255, 255, 255, 0.7)',
      disabled: 'rgba(255, 255, 255, 0.5)',
      hint: 'rgba(255, 255, 255, 0.5)',
    },
    light: {
      primary: 'rgba(0, 0, 0, 0.87)',
      secondary: 'rgba(0, 0, 0, 0.54)',
      disabled: 'rgba(0, 0, 0, 0.38)',
      hint: 'rgba(0, 0, 0, 0.38)',
    },
  },

  background: {
    dark: {
      default: stargateColors.midnightGreenEagleGreen,
      paper: stargateColors.blueSapphire,
    },
    light: {
      default: '#fafafa',
      paper: '#fff',
    },
  },
} as const;

export const colors = {
  /**
   * The palette is available to use, but prefer using the name.
   * This is only included for rare, one-off cases. Try to use
   * semantic names as much as possible.
   */
  palette,

  /**
   * A helper for making something see-thru.
   */
  transparent: 'rgba(0, 0, 0, 0)',

  derived,

  dark: {
    tint: palette.info100,
    border: derived.text.dark.primary,
    divider: 'rgba(255, 255, 255, 0.12)',
    separator: stargateColors.steelBlue,
    error: derived.error.main,
    errorBackground: derived.error.light,
    text: derived.text.dark,
    background: derived.background.dark,
  },
  light: {
    tint: stargateColors.blueGreen,
    border: derived.text.light.primary,
    divider: 'rgba(0, 0, 0, 0.12)',
    separator: stargateColors.steelBlue,
    error: derived.error.main,
    errorBackground: derived.error.light,
    text: derived.text.light,
    background: derived.background.light,
  },

  rnuilib: {
    colors: {
      bgDark: derived.background.dark.default,
      bgDarkPaper: derived.background.dark.paper,
      bgLight: derived.background.light.default,
      bgLightPaper: derived.background.light.paper,
      black: derived.text.light.primary,
      blacker: derived.text.light.secondary,
      blackish: derived.text.light.disabled,
      error: derived.error.main,
      info: derived.info.main,
      primary: derived.primary.main,
      secondary: derived.secondary.main,
      success: derived.success.main,
      warning: derived.warning.main,
      white: derived.text.dark.primary,
      whiter: derived.text.dark.secondary,
      whitish: derived.text.dark.disabled,
    },

    themes: {
      system: {},
      default: {},
      dark: {
        tint: palette.info100,
        border: derived.text.dark.primary,
        divider: 'rgba(255, 255, 255, 0.12)',
        separator: stargateColors.steelBlue,
        error: derived.error.main,
        errorBg: derived.error.light,
        text: derived.text.dark.primary,
        textSecondary: derived.text.dark.secondary,
        textDisabled: derived.text.dark.disabled,
        textHint: derived.text.dark.hint,
        bg: derived.background.dark.default,
        bgPaper: derived.background.dark.paper,
      },
      light: {
        tint: stargateColors.blueGreen,
        border: derived.text.light.primary,
        divider: 'rgba(0, 0, 0, 0.12)',
        separator: stargateColors.steelBlue,
        error: derived.error.main,
        errorBg: derived.error.light,
        text: derived.text.light.primary,
        textSecondary: derived.text.light.secondary,
        textDisabled: derived.text.light.disabled,
        textHint: derived.text.light.hint,
        bg: derived.background.light.default,
        bgPaper: derived.background.light.paper,
      },
    },
  },
} as const;
