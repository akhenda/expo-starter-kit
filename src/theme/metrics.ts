import { Dimensions, PixelRatio, Platform } from 'react-native';
import { getStatusBarHeight, ifIphoneX } from 'react-native-iphone-x-helper';

const { width, height } = Dimensions.get('window');

type DeviceSize = 'xsmall' | 'small' | 'normal' | 'large' | 'xlarge' | 'unknown' | 'default';

export function getDeviceSize(): DeviceSize {
  if (Platform.OS === 'ios') {
    // iPhone 4 spec
    if ((width <= 320 && height <= 480) || (width <= 480 && height <= 320)) return 'xsmall';

    // iPhone 5 spec
    if ((width <= 320 && height <= 568) || (width <= 568 && height <= 320)) return 'small';

    // iPhone 6 spec
    if ((width <= 375 && height <= 667) || (width <= 667 && height <= 375)) return 'normal';

    // iPhone 6 plus spec
    if ((width <= 414 && height <= 736) || (width <= 736 && height <= 414)) return 'large';

    // greater then iPhone 6 plus, such as tablets
    return 'xlarge';
  }

  if (Platform.OS === 'android') {
    // size spec from: https://developer.android.com/guide/practices/screens_support.html
    // xlarge screens are at least 960dp x 720dp
    // large screens are at least 640dp x 480dp
    // normal screens are at least 470dp x 320dp
    // small screens are at least 426dp x 320dp

    if ((width >= 960 && height >= 720) || (width >= 720 && width >= 960)) return 'xlarge';
    if ((width >= 640 && height >= 480) || (width >= 480 && height >= 640)) return 'large';
    if ((width >= 470 && height >= 320) || (width >= 320 && height >= 470)) return 'normal';
    if ((width >= 426 && height >= 320) || (width >= 320 && height >= 426)) return 'small';

    return 'xsmall';
  }

  return 'unknown';
}

export function selectDevice<T>(selection: { [size in DeviceSize]?: T } & { default: T }) {
  const size = getDeviceSize();
  const attr = selection[size];

  return attr || selection.default;
}

/**
 * A responsive font size helper adapted from https://stackoverflow.com/questions/33628677/react-native-responsive-font-size
 *
 * @param size The size unit in pixels
 */
export function rfs(size: number) {
  const scale = width / 350;
  const newSize = size * scale;

  if (Platform.OS === 'ios') return Math.round(PixelRatio.roundToNearestPixel(newSize));

  return Math.round(PixelRatio.roundToNearestPixel(newSize)) - 2;
}

/**
 * Global application metrics
 */
const metrics = {
  containerPadding: 15,
  deviceSize: getDeviceSize(),
  getStatusBarHeight,
  height: width < height ? height : width,
  ifIphoneX,
  isIphoneX: ifIphoneX,
  rfs,
  screenHeight: width < height ? height : width,
  screenWidth: width < height ? width : height,
  selectDevice,

  statusBarHeight: getStatusBarHeight(),
  tabBarHeight: 64,
  width: width < height ? width : height,
};

export default metrics;
