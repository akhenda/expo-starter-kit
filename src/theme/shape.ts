export const shape = {
  /**
   * The default border radius app-wide
   */
  borderRadius: 4,
} as const;
