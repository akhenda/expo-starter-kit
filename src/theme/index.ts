import { colors } from './colors';
import metrics, { rfs } from './metrics';
import styles from './styles';
import { typography } from './typography';

export * from './assets';
export * from './colors';
export * from './metrics';
export * from './spacing';
export * from './timing';
export * from './typography';

export const theme = {
  dark: {
    ...colors.dark,
    ...typography.dark,
    ...styles.dark,
  },
  light: {
    ...colors.light,
    ...typography.light,
    ...styles.light,
  },
  metrics,
  rfs,
};
