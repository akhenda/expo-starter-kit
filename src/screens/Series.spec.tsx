/**
 * Example React Native tests
 *
 * Ref:
 * @see https://github.com/vanGalilea/react-native-testing/blob/master/__tests__
 * @see https://testing-library.com/docs/react-native-testing-library/example-intro
 * @see https://testing-library.com/docs/react-native-testing-library/setup
 * @see https://callstack.github.io/react-native-testing-library/docs/react-navigation
 * @see https://testing-library.com/docs/react-testing-library/example-intro#full-example to
 * see an example that uses msw for api requests
 */
import { cleanup, fireEvent, screen } from '@testing-library/react-native';

import { typography } from '@src/theme';
import { render } from '@src/utils/tests';

import Series from './Series';

afterEach(cleanup);

it('renders correctly', async () => {
  const typedEmail = 'john@doe.me';

  render(<Series />);

  const theme = screen.getByTestId('theme');
  const roundButton = screen.getByTestId('button');
  const buttonText = screen.getByText(/change theme/i);
  const h1 = screen.getByText(/h1/i);
  const caption = screen.getByText(/caption/i);
  const overline = screen.getByText(/overline/i);

  expect(theme).toHaveTextContent('default');
  expect(h1).toHaveTextContent('H1 Series');
  expect(caption).toHaveTextContent('Caption');
  expect(overline).toHaveTextContent('Overline');
  expect(buttonText).toHaveTextContent('Change Theme');

  await fireEvent.changeText(screen.getByPlaceholderText(/Placeholder/i), typedEmail);

  // Using `findBy` query to wait for asynchronous operation to finish
  const email = await screen.findByTestId('email');

  // Using `toHaveTextContent` matcher from `@testing-library/jest-native` package.
  expect(email).toHaveTextContent(typedEmail);

  expect(h1).toHaveStyle({ color: 'rgba(0, 0, 0, 0.87)' });
  expect(h1).not.toHaveStyle({ color: '#fff' });

  fireEvent.press(buttonText);

  expect(theme).toHaveTextContent('dark');
  expect(h1).toHaveStyle({ ...typography.dark.h1 });
  expect(h1).toHaveStyle({ color: '#fff' });
  expect(h1).not.toHaveStyle({ color: 'rgba(0, 0, 0, 0.87)' });
  expect(buttonText.instance).toBeNull();

  fireEvent.press(roundButton);

  expect(theme).toHaveTextContent('light');
  expect(h1).toHaveStyle({ ...typography.light.h1 });
  expect(h1).toHaveStyle({ color: 'rgba(0, 0, 0, 0.87)' });
  expect(h1).not.toHaveStyle({ color: '#fff' });

  expect(screen.toJSON()).toMatchSnapshot();
});
