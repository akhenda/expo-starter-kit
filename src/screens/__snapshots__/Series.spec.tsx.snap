// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    [
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
      [
        {
          "position": "relative",
        },
        undefined,
        [
          {
            "paddingBottom": 0,
          },
        ],
      ],
    ]
  }
>
  <RCTScrollView>
    <View>
      <View
        padding-15={true}
        style={
          [
            undefined,
            undefined,
            {
              "flex": 1,
            },
            undefined,
            {
              "padding": 15,
            },
            {},
            {},
            undefined,
          ]
        }
      >
        <Text
          h1={true}
          style={
            [
              {
                "backgroundColor": "transparent",
                "color": "#20303C",
                "textAlign": "left",
              },
              {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "codaRegular",
                "fontSize": 206,
                "fontWeight": "300",
                "lineHeight": 240.40200000000002,
              },
              {
                "color": "rgba(0, 0, 0, 0.87)",
              },
              undefined,
              undefined,
              {},
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
          text={true}
        >
          H1 
          Series
        </Text>
        <Text
          h2={true}
          secondary={true}
          style={
            [
              {
                "backgroundColor": "transparent",
                "color": "#20303C",
                "textAlign": "left",
              },
              {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "codaRegular",
                "fontSize": 129,
                "fontWeight": "300",
                "lineHeight": 154.79999999999998,
              },
              {
                "color": "#57B21A",
              },
              undefined,
              undefined,
              {},
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
        >
          H2 
          Series
        </Text>
        <Text
          h3={true}
          style={
            [
              {
                "backgroundColor": "transparent",
                "color": "#20303C",
                "textAlign": "left",
              },
              {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "codaRegular",
                "fontSize": 103,
                "fontWeight": "400",
                "lineHeight": 120.20100000000001,
              },
              {
                "color": "#57B21A",
              },
              undefined,
              undefined,
              {},
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
          success={true}
        >
          H3 
          Series
        </Text>
        <Text
          h4={true}
          style={
            [
              {
                "backgroundColor": "transparent",
                "color": "#20303C",
                "textAlign": "left",
              },
              {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "codaRegular",
                "fontSize": 73,
                "fontWeight": "400",
                "lineHeight": 90.155,
              },
              {
                "color": "#fcc100",
              },
              undefined,
              undefined,
              {},
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
          warning={true}
        >
          H4 
          Series
        </Text>
        <Text
          h5={true}
          info={true}
          style={
            [
              {
                "backgroundColor": "transparent",
                "color": "#20303C",
                "textAlign": "left",
              },
              {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "codaRegular",
                "fontSize": 52,
                "fontWeight": "400",
                "lineHeight": 69.36800000000001,
              },
              {
                "color": "#2663ff",
              },
              undefined,
              undefined,
              {},
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
        >
          H5 
          Series
        </Text>
        <Text
          error={true}
          h6={true}
          style={
            [
              {
                "backgroundColor": "transparent",
                "color": "#20303C",
                "textAlign": "left",
              },
              {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "codaRegular",
                "fontSize": 43,
                "fontWeight": "500",
                "lineHeight": 68.8,
              },
              {
                "color": "#db4518",
              },
              undefined,
              undefined,
              {},
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
        >
          H6 
          Series
        </Text>
        <Text
          caption={true}
          style={
            [
              {
                "backgroundColor": "transparent",
                "color": "#20303C",
                "textAlign": "left",
              },
              {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "ubuntuRegular",
                "fontSize": 26,
                "fontWeight": "400",
                "lineHeight": 43.16,
              },
              undefined,
              undefined,
              undefined,
              {},
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
        >
          Caption
        </Text>
        <Text
          overline={true}
          style={
            [
              {
                "backgroundColor": "transparent",
                "color": "#20303C",
                "textAlign": "left",
              },
              {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "ubuntuRegular",
                "fontSize": 26,
                "fontWeight": "400",
                "lineHeight": 69.16,
                "textTransform": "uppercase",
              },
              undefined,
              undefined,
              undefined,
              {},
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
        >
          Overline
        </Text>
        <Text
          style={
            [
              {
                "backgroundColor": "transparent",
                "color": "#20303C",
                "textAlign": "left",
              },
              {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "ubuntuRegular",
                "fontSize": 35,
                "fontWeight": "400",
                "lineHeight": 61.25,
              },
              undefined,
              undefined,
              undefined,
              {},
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
          subtitle={true}
        >
          Subtitle
        </Text>
        <Text
          style={
            [
              {
                "backgroundColor": "transparent",
                "color": "#20303C",
                "textAlign": "left",
              },
              {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "ubuntuRegular",
                "fontSize": 30,
                "fontWeight": "500",
                "lineHeight": 47.1,
              },
              undefined,
              undefined,
              undefined,
              {},
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
          subtitle2={true}
        >
          Subtitle 2
        </Text>
        <Text
          body={true}
          style={
            [
              {
                "backgroundColor": "transparent",
                "color": "#20303C",
                "textAlign": "left",
              },
              {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "ubuntuRegular",
                "fontSize": 35,
                "fontWeight": "400",
                "lineHeight": 52.5,
              },
              undefined,
              undefined,
              undefined,
              {},
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
        >
          Body
        </Text>
        <Text
          body2={true}
          style={
            [
              {
                "backgroundColor": "transparent",
                "color": "#20303C",
                "textAlign": "left",
              },
              {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "ubuntuRegular",
                "fontSize": 30,
                "fontWeight": "400",
                "lineHeight": 42.9,
              },
              undefined,
              undefined,
              undefined,
              {},
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
          testID="email"
        >
          john@doe.me
        </Text>
        <Text
          p={true}
          style={
            [
              {
                "backgroundColor": "transparent",
                "color": "#20303C",
                "textAlign": "left",
              },
              {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "ubuntuRegular",
                "fontSize": 35,
                "fontWeight": "400",
                "lineHeight": 52.5,
              },
              undefined,
              undefined,
              undefined,
              {},
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
          testID="theme"
        >
          light
        </Text>
        <View
          style={
            [
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              [
                {},
                undefined,
                undefined,
                undefined,
              ],
            ]
          }
        >
          <View
            centerV={true}
            row={true}
            style={
              [
                undefined,
                undefined,
                undefined,
                undefined,
                {},
                {},
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                },
                [
                  {},
                  [
                    undefined,
                    undefined,
                  ],
                ],
              ]
            }
          >
            <View
              flexG={true}
              style={
                [
                  undefined,
                  undefined,
                  {
                    "flexGrow": 1,
                  },
                  undefined,
                  {},
                  {},
                  {},
                  undefined,
                ]
              }
            >
              <TextInput
                accessibilityState={
                  {
                    "disabled": false,
                  }
                }
                editable={true}
                maxLength={30}
                onBlur={[Function]}
                onChangeText={[Function]}
                onFocus={[Function]}
                placeholder="Placeholder"
                style={
                  [
                    {
                      "flexGrow": 1,
                      "paddingBottom": 0,
                      "paddingTop": 0,
                      "textAlign": "left",
                    },
                    {
                      "color": "#20303C",
                    },
                    [
                      {},
                      undefined,
                      undefined,
                      undefined,
                    ],
                    false,
                  ]
                }
                underlineColorAndroid="transparent"
                validate={
                  [
                    "required",
                    "email",
                    [Function],
                  ]
                }
                validationMessage={
                  [
                    "Field is required",
                    "Email is invalid",
                    "Password is too short",
                  ]
                }
                value="john@doe.me"
              />
            </View>
          </View>
          <View
            row={true}
            spread={true}
            style={
              [
                undefined,
                undefined,
                undefined,
                undefined,
                {},
                {},
                {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                },
                undefined,
              ]
            }
          >
            <Text
              $textDangerLight={true}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "color": "#20303C",
                    "textAlign": "left",
                  },
                  undefined,
                  {
                    "color": "#FC3D2F",
                  },
                  undefined,
                  undefined,
                  {},
                  undefined,
                  undefined,
                  undefined,
                  [
                    {
                      "flexShrink": 1,
                      "minHeight": 20,
                    },
                    undefined,
                  ],
                ]
              }
              testID="undefined.validationMessage"
            />
            <Text
              $textNeutral={true}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "color": "#20303C",
                    "textAlign": "left",
                  },
                  undefined,
                  {
                    "color": "#6E7881",
                  },
                  undefined,
                  undefined,
                  {},
                  undefined,
                  undefined,
                  undefined,
                  [
                    {
                      "flexGrow": 1,
                      "textAlign": "right",
                    },
                    undefined,
                  ],
                ]
              }
              testID="undefined.charCounter"
            >
              11/30
            </Text>
          </View>
        </View>
        <View
          accessibilityRole="button"
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onLayout={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "alignSelf": "flex-end",
              "backgroundColor": "#1C7AB0",
              "borderRadius": 999,
              "elevation": 2,
              "flexDirection": "row",
              "justifyContent": "center",
              "marginTop": 15,
              "minWidth": 90,
              "opacity": 1,
              "paddingHorizontal": 20,
              "paddingVertical": 9.5,
              "shadowColor": "#1C7AB0",
              "shadowOffset": {
                "height": 5,
                "width": 0,
              },
              "shadowOpacity": 0.35,
              "shadowRadius": 9.5,
            }
          }
          testID="button"
        >
          <Text
            numberOfLines={1}
            style={
              [
                {
                  "backgroundColor": "transparent",
                  "color": "#20303C",
                  "textAlign": "left",
                },
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                [
                  {
                    "backgroundColor": "transparent",
                    "flexDirection": "row",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": "400",
                    "lineHeight": 24,
                  },
                  {
                    "color": "#FFFFFF",
                  },
                  undefined,
                  undefined,
                  undefined,
                ],
              ]
            }
            testID="button.label"
          >
            Change Theme
          </Text>
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "#D3E4FF",
                  "fontSize": 24,
                },
                [
                  {
                    "marginLeft": undefined,
                    "marginRight": undefined,
                  },
                  undefined,
                ],
                {
                  "fontFamily": "ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
